### ISSUE TITLE
[BUG] Possible segfault in exception handling code

### ISSUE THREAD

**Issue by kulikjak** [cython/cython#4722]
**Describe the bug**
We found out a possible segfault in the exception handling code of [lxml](https://pypi.org/project/lxml/) 4.7.1 generated with Cython (running Python 3.7).

The problematic code generated with Cython:
```C
__pyx_t_6 = __Pyx_PyErr_ExceptionMatches(__pyx_builtin_ImportError);
if (__pyx_t_6) {
  __Pyx_AddTraceback("lxml.etree", __pyx_clineno, __pyx_lineno,
__pyx_filename);
  if (__Pyx_GetException(&__pyx_t_15, &__pyx_t_14, &__pyx_t_1) < 0) ...
```

The issue is that between the condition check and the `__Pyx_GetException()` call, the exception can in some cases be reset in `__Pyx_AddTraceback()` function.

The complete call trace is here (Python 3.7):

_lxml_ - etree.c
-> `__Pyx_AddTraceback()` -> `__Pyx_CreateCodeObjectForTraceback()`
_CPython_ - codeobject.c
-> [`PyCode_NewEmpty()`](https://github.com/python/cpython/blob/387f93c156288c170ff0016a75af06e109d48ee1/Objects/codeobject.c#L245)
_CPython_ - unicodeobject.c
-> [`PyUnicode_DecodeFSDefault()`](https://github.com/python/cpython/blob/387f93c156288c170ff0016a75af06e109d48ee1/Objects/unicodeobject.c#L3666) -> [`PyUnicode_DecodeFSDefaultAndSize()`](https://github.com/python/cpython/blob/387f93c156288c170ff0016a75af06e109d48ee1/Objects/unicodeobject.c#L3686) -> [`PyUnicode_Decode()`](https://github.com/python/cpython/blob/387f93c156288c170ff0016a75af06e109d48ee1/Objects/unicodeobject.c#L3237)
_CPython_ - codecs.c
-> [`_PyCodec_DecodeText()`](https://github.com/python/cpython/blob/387f93c156288c170ff0016a75af06e109d48ee1/Python/codecs.c#L618) -> [`_PyCodec_DecodeInternal()`](https://github.com/python/cpython/blob/387f93c156288c170ff0016a75af06e109d48ee1/Python/codecs.c#L471)
_CPython_ - call.c
-> [`PyEval_CallObject()`](https://github.com/python/cpython/blob/387f93c156288c170ff0016a75af06e109d48ee1/Objects/call.c#L802) - this should not be called with active exception (as the commend there says).

The execution takes a slightly different path in our case when using Python 3.9, but the 'exception clearing functions' can still be called in some circumstances.

**To Reproduce**
Code to reproduce the behaviour:

We hit this only when using specific locale setting `LANG="en_US.ISO8859-15"` (so that additional decoding is necessary). The interpreter doesn't always segfault, but running with gdb
```shell
> LC_ALL="" LANG="en_US.ISO8859-15"  gdb --args /usr/bin/python3.7 -Esc 'from lxml import etree'
```
```gdb
b __Pyx_AddTraceback
y
r
b PyEval_CallObjectWithKeywords
b __Pyx__GetException
c
```
clearly shows that in between `__Pyx_AddTraceback()` and `__Pyx__GetException()` breakpoints, `PyEval_CallObjectWithKeywords()` is hit.

**Expected behavior**
No segfault - I believe that the exception should be saved and restored after the problematic call.

**Environment (please complete the following information):**
 - OS: Oracle Solaris 11.4 (but I do believe this can be hit by other OSes as well)
 - Python version 3.7 (but can be hit in 3.9 as well)
 - The same code is generated with Cython 29.21 (used to build lxml 4.7.1, I believe), Cython 29.28 and 3.0.0a10.

**Possible fix**
After using the following patch on lxml, the segfault is gone:
```C
--- lxml-4.7.1/src/lxml/etree.c
+++ lxml-4.7.1/src/lxml/etree.c
@@ -267792,8 +267792,14 @@ static void __Pyx_AddTraceback(const cha
     }
     py_code = __pyx_find_code_object(c_line ? -c_line : py_line);
     if (!py_code) {
+        /* __Pyx_CreateCodeObjectForTraceback can in some cases subsequently
+           call functions that cannot be called with set exceptions */
+        PyObject *etype, *eval, *etb;
+        PyErr_Fetch(&etype, &eval, &etb);
         py_code = __Pyx_CreateCodeObjectForTraceback(
             funcname, c_line, py_line, filename);
+        PyErr_Restore(etype, eval, etb);
+
         if (!py_code) goto bad;
         __pyx_insert_code_object(c_line ? -c_line : py_line, py_code);
     }
```

**Comment by scoder** [comment]
Thanks for investigating this. Yes, this is an issue. And the fix you provide also seems ok. There is no performance issue here since we should normally have a cached code object already.

However, if the code object creation fails, then we shouldn't reset the exception here but clear the fetched exception references and propagate the new exception (`goto bad`). This will mostly likely only happen with memory allocation issues.

**Comment by kulikjak** [comment]
I see, makes sense. I created a fix via #4723
